#101 Add requirements for extension context (#124)
exasol/extension-manager: doc/design.md

Needs: impl, utest, itest

### Extension Context

> The extension context allows extension definitions to interact with the database and extension manager e.g. by executing queries.

Reviewer: ```suggestion
The extension context allows extension definitions to interact with the extension manager and the database (e.g. by executing queries).
```

#### Extension Context SQL Client
`dsn~extension-context-sql-client~1`

> The SQL client in the extension context allows extension definitions to execute statements and run queries against the database and process query results.

Reviewer: ```suggestion
The SQL client in the extension context allows the extension definition to execute statements and run queries against the database and process query results.
```

Rationale:

> * This allows extensions to create necessary database objects like `SCRIPT`s, `CONNECTION`s and `VIRTUAL SCHEMA`s.
> * Extensions can also run arbitrary queries in order to read required information about their installations (e.g. scripts) and instances (e.g. virtual schemas).
> * While EM also provides [access to metadata](#extension-context-metadata) via the context, this information may not be sufficient. Executing arbitrary queries ensures maximum flexibility for extensions.

Reviewer: ```suggestion
* This allows the extension to create necessary database objects like `SCRIPT`s, `CONNECTION`s and `VIRTUAL SCHEMA`s.
* The extension can also run arbitrary queries in order to read required information about their installations (e.g. scripts) and instances (e.g. virtual schemas).
* While EM also provides [access to metadata](#extension-context-metadata) via the context, this information may not be sufficient. Executing arbitrary queries ensures maximum flexibility for extensions.
```

Covers:
* [`req~install-extension-database-objects~1`](system_requirements.md#update-extension)

Needs: impl, utest, itest

#### Extension Context BucketFS
`dsn~extension-context-bucketfs~1`

> The BucketFS client in the extension context allows extension definitions to interact with BucketFS.

Reviewer: ```suggestion
The BucketFS client in the extension context allows the extension definition to interact with BucketFS.
```

Covers:
* [`req~install-extension-database-objects~1`](system_requirements.md#update-extension)

Needs: impl, utest, itest

##### Resolving Files in BucketFS
`dsn~resolving-files-in-bucketfs~1`

EM resolves filenames to absolute paths in BucketFS.

Rationale:

Extensions require files in BucketFS. The exact location in BucketFS of the files is not known when creating an extension, only the file name (e.g. `document-files-virtual-schema-dist-7.3.3-s3-2.6.2.jar`). In order to create adapter scripts, the extension must know the absolute path of the file in BucketFS (e.g. `/buckets/bfsdefault/default/document-files-virtual-schema-dist-7.3.3-s3-2.6.2.jar`). So EM must allow the extension to resolve file names to an absolute path in BucketFS.

Needs: impl, utest, itest

#### Extension Context Metadata
`dsn~extension-context-metadata~1`

> The Metadata client in the extension context allows extension definitions to read Metadata (e.g. tables `SYS.EXA_ALL_SCRIPTS` or `SYS.EXA_ALL_VIRTUAL_SCHEMAS`) from the database.

Reviewer: ```suggestion
The Metadata client in the extension context allows the extension definition to read Metadata (e.g. tables `SYS.EXA_ALL_SCRIPTS` or `SYS.EXA_ALL_VIRTUAL_SCHEMAS`) from the database.
```

Rationale:

> * This information is necessary for extensions to find installations (e.g. scripts) and instances (e.g. virtual schemas).

Reviewer: ```suggestion
* This information is necessary for the extension to find its installations (e.g. scripts) and instances (e.g. virtual schemas).
```
